### PR TITLE
[ci]: Fix auto-labeling

### LINKED ISSUE #4278: Automatic notification about config.json changes through CI
As we discussed with Vasily, it'll be much more convenient if he's notified about any possible configuration change.
Ultimately, we concluded that it's best to have a CI utility that doesn't use any outside software and notifies him about the changes.
It's not a priority and probably will be added in March 2024.

### PR DESCRIPTION
**Pull request by s8sato**
## Description

- Fix: `api-changes` `config-changes` auto-labeling
- Refactor: merge the above and `iroha2` into one workflow
- Feature: notify someone specific based on the above auto-labels
- Feature: add another group of auto-labels to categorize PRs based on title

### Linked issue

- Closes #4278

### Benefits

- Make PR labels easier to handle
- Reduce workflows by one

### Checklist

- [x] I've read `CONTRIBUTING.md`
- [x] I've used the standard signed-off commit format (or will squash just before merging)
- [x] All applicable CI checks pass (or I promised to make them pass later)
- [ ] (optional) I've written unit tests for the code changes
- [ ] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

File: .github/workflows/iroha2-dev-pr-label.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev-pr-label.yml:33]: image
https://docs.github.com/en/actions/learn-github-actions/expressions#contains

File: .github/workflows/iroha2-dev-pr-label.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev-pr-label.yml:8]: image
https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#patterns-to-match-branches-and-tags

**Comment by 0x009922** [comment]
Can we wait with merging of this until the merge of #4239? That PR changes the approach to the config itself. Actually, after the 2.0 release there should NOT be breaking changes in the config and we should establish a clear deprecation and migration policy.

**Comment by BAStos525** [comment]
From my side, it would be better to be notified via PR auto-assignment or just auto-comment with my username because it's better rather than to search for PRs with `config-changed` label. Smth like that:
```
permissions:
  pull-requests: write
jobs:
  api-changes:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: jitterbit/get-changed-files@v1
        continue-on-error: true
        id: api_label
      - uses: actions-ecosystem/action-add-labels@v1
        if: contains(steps.api_label.outputs.added_modified, 'docs/source/references/schema.json')
        with:
          github_token: ${{ secrets.github_token }}
          labels: |
            api-changes
      - uses: mshick/add-pr-comment@v2
        if: contains(steps.api_label.outputs.added_modified, 'docs/source/references/schema.json')
        with:
          message: @BAStos525 
```

But I understand that we should'n bind CI within any particular person.

**Comment by s8sato** [comment]
https://github.com/hyperledger/iroha/pull/4276#issuecomment-1942049018
Yeah, for now this PR is only intended to add clues to detect configuration changes.
Here are 2 other suggestions for making notifications actually work.
Do you think the followings would work as well as your suggestion @BAStos525  ?

1. Have a separate workflow dedicated to managing individual notifications for labeling, including manual labeling
   - https://github.com/marketplace/actions/issue-label-notifier
2. Have each person have web hooks to receive notifications
   - https://github.com/bcongdon/github-label-notify

But before that, we have yet to see CI label `config-changes` in response to mock config changes in this PR

**Comment by BAStos525** [comment]
> [#4276 (comment)](https://github.com/hyperledger/iroha/pull/4276#issuecomment-1942049018)
> 
> Yeah, for now this PR is only intended to add clues to detect configuration changes. Here are 2 other suggestions for making notifications actually work. Do you think the followings would work as well as your suggestion @BAStos525 ?
> 
> 1. Have a separate workflow dedicated to managing individual notifications for labeling, including manual labeling
>    
>    * https://github.com/marketplace/actions/issue-label-notifier
> 2. Have each person have web hooks to receive notifications
>    
>    * https://github.com/bcongdon/github-label-notify
> 
> But before that, we have yet to see CI label `config-changes` in response to mock config changes in this PR

We will think about it later. I want say that we should try to avoid creating a new workflows for tiny tasks whatever it's possible.

File: .github/workflows/iroha2-dev-pr-title.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev-pr-title.yml:4]: to gain write access to PRs

File: .github/workflows/iroha2-label.yml
Comment (s8sato) [inline .github/workflows/iroha2-label.yml:14]: https://github.com/actions/labeler

I think this is a good consolidation of `jitterbit/get-changed-files` and `actions-ecosystem/action-add-labels`

**Review by s8sato — COMMENTED** [review]
This PR itself cannot trigger the changed workflows since `pull_request_target` event-driven workflow is based on the code of the base branch (`iroha2-dev`)
https://github.com/actions/labeler?tab=readme-ov-file#initial-set-up-of-the-labeler-action

File: .github/labeler.yml
Comment (0x009922) [inline .github/labeler.yml:8]: And, perhaps, `iroha_torii` and `iroha_torii_const` crates sources?

**Review by 0x009922 — APPROVED** [review]

File: .github/labeler.yml
Comment (BAStos525) [inline .github/labeler.yml:1]: Is this file connected to `I2::Label` workflow?

Comment (s8sato) [reply .github/labeler.yml:1]: Yes, this is the configuration file for `actions/labeler`. See https://github.com/actions/labeler#create-githublabeleryml

**Review by BAStos525 — DISMISSED** [review]

File: .github/workflows/iroha2-dev-pr.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev-pr.yml:12]: to not trigger unnecessary workflows after this PR

File: .github/labeler.yml
Comment (s8sato) [inline .github/labeler.yml:11]: https://github.com/hyperledger/iroha/pull/4276#discussion_r1495376696 reflected, please re-approve @0x009922  @BAStos525

**Review by BAStos525 — APPROVED** [review]

File: .github/workflows/iroha2-dev-pr.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev-pr.yml:10]: I don't think we need to trigger tests by changing any workflow except for the testing workflow itself. What do you think @mversic

**Review by 0x009922 — APPROVED** [review]

File: .github/workflows/iroha2-dev-pr.yml
Comment (mversic) [reply .github/workflows/iroha2-dev-pr.yml:10]: agreed

**Pull request merged**
